File: app/src/main/java/com/example/nonameapp/base/BaseAdapter.java
package com.example.nonameapp.base;

import android.view.LayoutInflater;
import android.view.ViewGroup;

import androidx.recyclerview.widget.RecyclerView;
import androidx.viewbinding.ViewBinding;

import java.util.ArrayList;
import java.util.List;

public abstract class BaseAdapter<T, VB extends ViewBinding> extends RecyclerView.Adapter<BaseViewHolder<VB>> {
    private final BindingInflater<VB> bindingInflater;
    private final List<T> dataList = new ArrayList<>();

    protected BaseAdapter(BindingInflater<VB> bindingInflater) {
        this.bindingInflater = bindingInflater;
    }

    @Override
    public BaseViewHolder<VB> onCreateViewHolder(ViewGroup parent, int viewType) {
        VB binding = bindingInflater.inflate(LayoutInflater.from(parent.getContext()));
        return new BaseViewHolder<>(binding);
    }

    @Override
    public int getItemCount() {
        return dataList.size();
    }

    @Override
    public void onBindViewHolder(BaseViewHolder<VB> holder, int position) {
        bindData(holder.getBinding(), dataList.get(position), position);
        onItemClick(holder.getBinding(), dataList.get(position), position);
    }

    protected abstract void bindData(VB binding, T item, int position);

    protected abstract void onItemClick(VB binding, T item, int position);

    public void setData(int position, T data) {
        if (position >= dataList.size()) {
            return;
        }
        dataList.set(position, data);
        notifyItemChanged(position);
    }

    public void removeData(int position) {
        if (position >= dataList.size()) {
            return;
        }
        dataList.remove(position);
        notifyItemRemoved(position);
    }

    public interface BindingInflater<VB extends ViewBinding> {
        VB inflate(LayoutInflater inflater);
    }
}

File: app/src/main/java/com/example/nonameapp/base/BaseAdapter.kt
package com.example.nonameapp.base

import android.view.LayoutInflater
import android.view.ViewGroup
import androidx.recyclerview.widget.RecyclerView.Adapter
import androidx.viewbinding.ViewBinding

abstract class BaseAdapter<T, VB : ViewBinding>(
    private val bindingInflater: (LayoutInflater, ViewGroup, Boolean) -> VB,
    protected var dataList: MutableList<T> = mutableListOf()
) : Adapter<BaseViewHolder<VB>>() {
    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): BaseViewHolder<VB> {
        val binding = bindingInflater(LayoutInflater.from(parent.context), parent, false)
        return BaseViewHolder(binding = binding)
    }

    override fun getItemCount(): Int = dataList.size

    override fun onBindViewHolder(holder: BaseViewHolder<VB>, position: Int) {
        bindData(holder.binding, dataList[position], position)
        onItemClick(holder.binding, dataList[position], position)
    }

    abstract fun bindData(binding: VB, item: T, position: Int)

    abstract fun onItemClick(binding: VB, item: T, position: Int)

    fun setData(position: Int, data: T) {
        if (position >= dataList.size) {
            return
        }
        dataList[position] = data
        notifyItemChanged(position)
    }

    fun removeData(position: Int) {
        if (position >= dataList.size) {
            return
        }
        this.dataList.removeAt(position)
        notifyItemRemoved(position)
    }
}